src/bfd-disasm.cpp

```diff
@@ -0,0 +1,121 @@
+#include <iostream>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <bfd.h>
+#include <dis-asm.h>
+#include "bcc_syms.h"
+#include "bcc_elf.h"
+#include "bfd-disasm.h"
+
+namespace bpftrace {
+
+BfdDisasm::BfdDisasm(std::string &path)
+{
+  fd_ = open(path.c_str(), O_RDONLY);
+}
+
+BfdDisasm::~BfdDisasm()
+{
+  if (fd_ >= 0)
+    close(fd_);
+}
+
+static void get_exec_path(char *tpath, size_t size)
+{
+  const char *path = "/proc/self/exe";
+  ssize_t len;
+
+  len = readlink(path, tpath, size - 1);
+  tpath[len] = 0;
+}
+
+static int fprintf_nop(void *out __attribute__((unused)), const char *fmt __attribute__((unused)), ...)
+{
+  return 0;
+}
+
+static AlignState is_aligned_buf(void *buf, uint64_t size, uint64_t offset)
+{
+  disassembler_ftype disassemble;
+  struct disassemble_info info;
+  char tpath[4096];
+  bfd *bfdf;
+
+  get_exec_path(tpath, sizeof(tpath));
+
+  bfdf = bfd_openr(tpath, NULL);
+  if (bfdf == NULL)
+    return AlignState::Fail;
+
+  if (!bfd_check_format(bfdf, bfd_object))
+  {
+    bfd_close(bfdf);
+    return AlignState::Fail;
+  }
+
+  init_disassemble_info(&info, stdout, fprintf_nop);
+
+  info.arch = bfd_get_arch(bfdf);
+  info.mach = bfd_get_mach(bfdf);
+  info.buffer = static_cast<bfd_byte*>(buf);
+  info.buffer_length = size;
+
+  disassemble_init_for_target(&info);
+
+#ifdef LIBBFD_DISASM_FOUR_ARGS_SIGNATURE
+  disassemble = disassembler(info.arch,
+           bfd_big_endian(bfdf),
+           info.mach,
+           bfdf);
+#else
+  disassemble = disassembler(bfdf);
+#endif
+
+  uint64_t pc = 0;
+  int count;
+
+  do {
+    count = disassemble(pc, &info);
+    pc += static_cast<uint64_t>(count);
+
+    if (pc == offset)
+    {
+      bfd_close(bfdf);
+      return AlignState::Ok;
+    }
+
+  } while (static_cast<uint64_t>(count) > 0 && pc < size && pc < offset);
+
+  bfd_close(bfdf);
+  return AlignState::NotAlign;
+}
+
+AlignState BfdDisasm::is_aligned(uint64_t offset, uint64_t pc)
+{
+  AlignState aligned = AlignState::Fail;
+  // 100 bytes should be enough to cover next instruction behind pc
+  uint64_t size = pc + 100;
+  void *buf;
+
+  if (fd_ < 0)
+    return aligned;
+
+  buf = malloc(size);
+  if (!buf)
```

Comment:
I think it would be good to notify the user when a malloc fails.

Reply:
I noticed few `perror` calls in code, would that be enough? like:
`  perror("malloc failed")`

or rather something like:
`  std::cerr << "memory allocation failed" << std::endl;`

Reply:
The perror one is fine imo, the errno value can be useful.

Should this be fatal too? With unsafe it would keep going. Although I'd expect it to crash in some other place.

Reply:
if it's bigger issue, we'll crash at some point I guess.. if it's minor flux
we allow it to continue, because at this point the allocation is not crucial
for the rest of the run.. so, I'd keep it going